### DISCUSSION TITLE
Want to print info which are offloaded to Device(GPU)

### DISCUSSION THREAD

**Discussion by devAj10** [pocl/pocl#1419] [Q&A] [answered]
Hii,
I would like to print the last items that are transferred to the Device for executing a Kernel.
I know the Data structure ( __kernel_exec_cmd   and  __kernel_metada ). I just want to print just before it's offloaded to Device.
Please let me know if it's possible to print or not.

**Comment by pjaaskel** [comment]
Assuming you are talking about the CPU/pthread device. The work-group function is called here: https://github.com/pocl/pocl/blob/main/lib/CL/devices/pthread/pthread_scheduler.c#L328, perhaps around there you can add your debug printouts.

**Reply by devAj10** [reply]
Sorry for the Wrong title, I was talking about GPUs or Accelerators.
Let's take an eg. Kernel for vec sum

`kernel void
vecadd (__global const float *a,
	__global const float *b,
	__global float *c)
{
  int gid = get_global_id(0);
  c[gid] = a[gid] + b[gid];
}`

So my question is How is this offloaded to the device? is it related to pocl_device.h file? How can we print?

**Reply by pjaaskel** [reply, marked as answer]
The different "drivers" or backends are located in lib/CL/devices, for an example of a non-CPU (GPU device), you can check the level zero, cuda or the LZ drivers for hints on how to create a new driver for a new accelerator device. This is also a good starting doc: http://portablecl.org/docs/html/driver_design.html
